File: frontend/src/types/configManagement.ts
/**
 * Configuration Management Types
 * Advanced Configuration Management System for Chronos AI Agent Builder Studio
 */

// ============== Enum Types ==============

export type SchemaType = 'AGENT' | 'WORKFLOW' | 'SYSTEM' | 'CUSTOM';

export type SnapshotType = 'MANUAL' | 'AUTOMATED' | 'SCHEDULED';

export type EnvironmentType = 'DEVELOPMENT' | 'STAGING' | 'PRODUCTION';

export type ChangeType = 'ADDED' | 'REMOVED' | 'MODIFIED';

// ============== ConfigSchema Types ==============

export interface ConfigSchema {
  id: string;
  name: string;
  description?: string;
  schema_type: SchemaType;
  schema_definition: Record<string, unknown>;
  version: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ConfigSchemaCreate {
  name: string;
  description?: string;
  schema_type: SchemaType;
  schema_definition: Record<string, unknown>;
  is_active?: boolean;
}

export interface ConfigSchemaUpdate {
  name?: string;
  description?: string;
  schema_type?: SchemaType;
  schema_definition?: Record<string, unknown>;
  is_active?: boolean;
}

// ============== ConfigVersion Types ==============

export interface ConfigVersion {
  id: string;
  config_id?: string;
  config_schema_id: string;
  version_number: number;
  configuration_data: Record<string, unknown>;
  changelog?: string;
  created_by?: string;
  created_at: string;
}

export interface ConfigVersionCreate {
  config_id?: string;
  config_schema_id: string;
  configuration_data: Record<string, unknown>;
  changelog?: string;
  created_by?: string;
}

// ============== ConfigSnapshot Types ==============

export interface ConfigSnapshot {
  id: string;
  config_id: string;
  version_id: string;
  snapshot_data: Record<string, unknown>;
  snapshot_type: SnapshotType;
  created_at: string;
}

export interface ConfigSnapshotCreate {
  config_id: string;
  version_id: string;
  snapshot_data: Record<string, unknown>;
  snapshot_type?: SnapshotType;
}

// ============== ConfigTemplate Types ==============

export interface ConfigTemplate {
  id: string;
  name: string;
  description?: string;
  config_schema_id: string;
  template_data: Record<string, unknown>;
  parameters?: Record<string, unknown>;
  is_public: boolean;
  user_id?: string;
  created_at: string;
  updated_at: string;
}

export interface ConfigTemplateCreate {
  name: string;
  description?: string;
  config_schema_id: string;
  template_data: Record<string, unknown>;
  parameters?: Record<string, unknown>;
  is_public?: boolean;
  user_id?: string;
}

export interface ConfigTemplateUpdate {
  name?: string;
  description?: string;
  template_data?: Record<string, unknown>;
  parameters?: Record<string, unknown>;
  is_public?: boolean;
}

// ============== EnvironmentConfig Types ==============

export interface EnvironmentConfig {
  id: string;
  config_id: string;
  environment: EnvironmentType;
  config_data: Record<string, unknown>;
  overrides?: Record<string, unknown>;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface EnvironmentConfigCreate {
  config_id: string;
  environment: EnvironmentType;
  config_data: Record<string, unknown>;
  overrides?: Record<string, unknown>;
  is_active?: boolean;
}

export interface EnvironmentConfigUpdate {
  config_data?: Record<string, unknown>;
  overrides?: Record<string, unknown>;
  is_active?: boolean;
}

// ============== Configuration Types ==============

export interface Configuration {
  id: string;
  name: string;
  description?: string;
  config_schema_id: string;
  current_version_id?: string;
  is_active: boolean;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

export interface ConfigurationCreate {
  name: string;
  description?: string;
  config_schema_id: string;
  created_by?: string;
}

export interface ConfigurationUpdate {
  name?: string;
  description?: string;
  current_version_id?: string;
  is_active?: boolean;
}

// ============== Validation Types ==============

export interface ValidationError {
  path: string;
  message: string;
  error_type: string;
}

export interface ConfigValidationResponse {
  is_valid: boolean;
  errors: ValidationError[];
  schema_id: string;
  validated_at: string;
}

export interface ConfigValidationRequest {
  config_data: Record<string, unknown>;
  schema_id: string;
}

// ============== Diff Types ==============

export interface DiffEntry {
  path: string;
  old_value: unknown;
  new_value: unknown;
  change_type: ChangeType;
}

export interface ConfigDiffResponse {
  has_differences: boolean;
  differences: DiffEntry[];
  added_count: number;
  removed_count: number;
  modified_count: number;
  compared_at: string;
}

export interface ConfigDiffRequest {
  config_a: Record<string, unknown>;
  config_b: Record<string, unknown>;
}

// ============== Rollback Types ==============

export interface RollbackRequest {
  version_id: string;
  changelog?: string;
}

export interface RollbackResponse {
  success: boolean;
  previous_version?: ConfigVersion;
  new_version?: ConfigVersion;
  message: string;
}

// ============== Template Application Types ==============

export interface ApplyTemplateRequest {
  template_id: string;
  parameters?: Record<string, unknown>;
}

export interface ApplyTemplateResponse {
  success: boolean;
  applied_config: Record<string, unknown>;
  template_id: string;
  applied_at: string;
}

// ============== Paginated Response Types ==============

export interface PaginatedConfigSchemaResponse {
  items: ConfigSchema[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface PaginatedConfigVersionResponse {
  items: ConfigVersion[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface PaginatedConfigTemplateResponse {
  items: ConfigTemplate[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface PaginatedConfigSnapshotResponse {
  items: ConfigSnapshot[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

// ============== UI State Types ==============

export interface ConfigEditorState {
  currentConfig: Record<string, unknown> | null;
  selectedSchemaId: string | null;
  selectedVersionId: string | null;
  selectedEnvironment: EnvironmentType;
  isDirty: boolean;
  isValidating: boolean;
  validationErrors: ValidationError[];
}

export interface DiffViewerState {
  configA: Record<string, unknown> | null;
  configB: Record<string, unknown> | null;
  differences: DiffEntry[];
  filter: ChangeType | 'ALL';
}

export interface TemplateGalleryState {
  templates: ConfigTemplate[];
  selectedTemplate: ConfigTemplate | null;
  previewParameters: Record<string, unknown>;
  isApplying: boolean;
}

// ============== JSON Schema Types ==============

export interface JSONSchema {
  type?: string;
  properties?: Record<string, JSONSchema>;
  required?: string[];
  description?: string;
  enum?: unknown[];
  default?: unknown;
  items?: JSONSchema;
  additionalProperties?: boolean | JSONSchema;
}

File: backend/app/schemas/config_management.py
"""
Pydantic Schemas for Configuration Management
Advanced Configuration Management System for Chronos AI Agent Builder Studio
"""

import enum
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from uuid import UUID


# Enums matching database models
class SchemaType(str, enum.Enum):
    AGENT = "AGENT"
    WORKFLOW = "WORKFLOW"
    SYSTEM = "SYSTEM"
    CUSTOM = "CUSTOM"


class SnapshotType(str, enum.Enum):
    MANUAL = "MANUAL"
    AUTOMATED = "AUTOMATED"
    SCHEDULED = "SCHEDULED"


class EnvironmentType(str, enum.Enum):
    DEVELOPMENT = "DEVELOPMENT"
    STAGING = "STAGING"
    PRODUCTION = "PRODUCTION"


# ============== ConfigSchema Schemas ==============

class ConfigSchemaBase(BaseModel):
    """Base schema for ConfigSchema"""
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    schema_type: SchemaType = SchemaType.CUSTOM
    schema_definition: Dict[str, Any]


class ConfigSchemaCreate(ConfigSchemaBase):
    """Schema for creating a new ConfigSchema"""
    is_active: Optional[bool] = True


class ConfigSchemaUpdate(BaseModel):
    """Schema for updating a ConfigSchema"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    schema_type: Optional[SchemaType] = None
    schema_definition: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None


class ConfigSchemaResponse(ConfigSchemaBase):
    """Schema for ConfigSchema response"""
    id: UUID
    version: int
    is_active: bool
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ============== ConfigVersion Schemas ==============

class ConfigVersionBase(BaseModel):
    """Base schema for ConfigVersion"""
    config_schema_id: UUID
    configuration_data: Dict[str, Any]
    changelog: Optional[str] = None


class ConfigVersionCreate(ConfigVersionBase):
    """Schema for creating a new ConfigVersion"""
    config_id: Optional[UUID] = None
    created_by: Optional[str] = None


class ConfigVersionResponse(BaseModel):
    """Schema for ConfigVersion response"""
    id: UUID
    config_id: Optional[UUID]
    config_schema_id: UUID
    version_number: int
    configuration_data: Dict[str, Any]
    changelog: Optional[str]
    created_by: Optional[str]
    created_at: datetime
    
    class Config:
        from_attributes = True


class ConfigVersionWithSchema(ConfigVersionResponse):
    """Schema for ConfigVersion with included schema info"""
    config_schema: Optional[ConfigSchemaResponse] = None


# ============== ConfigSnapshot Schemas ==============

class ConfigSnapshotBase(BaseModel):
    """Base schema for ConfigSnapshot"""
    config_id: UUID
    version_id: UUID
    snapshot_data: Dict[str, Any]
    snapshot_type: SnapshotType = SnapshotType.MANUAL


class ConfigSnapshotCreate(ConfigSnapshotBase):
    """Schema for creating a new ConfigSnapshot"""
    pass


class ConfigSnapshotResponse(BaseModel):
    """Schema for ConfigSnapshot response"""
    id: UUID
    config_id: UUID
    version_id: UUID
    snapshot_data: Dict[str, Any]
    snapshot_type: SnapshotType
    created_at: datetime
    
    class Config:
        from_attributes = True


# ============== ConfigTemplate Schemas ==============

class ConfigTemplateBase(BaseModel):
    """Base schema for ConfigTemplate"""
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    config_schema_id: UUID
    template_data: Dict[str, Any]
    parameters: Optional[Dict[str, Any]] = None


class ConfigTemplateCreate(ConfigTemplateBase):
    """Schema for creating a new ConfigTemplate"""
    is_public: Optional[bool] = False
    user_id: Optional[UUID] = None


class ConfigTemplateUpdate(BaseModel):
    """Schema for updating a ConfigTemplate"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    template_data: Optional[Dict[str, Any]] = None
    parameters: Optional[Dict[str, Any]] = None
    is_public: Optional[bool] = None


class ConfigTemplateResponse(ConfigTemplateBase):
    """Schema for ConfigTemplate response"""
    id: UUID
    is_public: bool
    user_id: Optional[UUID]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ============== EnvironmentConfig Schemas ==============

class EnvironmentConfigBase(BaseModel):
    """Base schema for EnvironmentConfig"""
    config_id: UUID
    environment: EnvironmentType
    config_data: Dict[str, Any]
    overrides: Optional[Dict[str, Any]] = None


class EnvironmentConfigCreate(EnvironmentConfigBase):
    """Schema for creating a new EnvironmentConfig"""
    is_active: Optional[bool] = True


class EnvironmentConfigUpdate(BaseModel):
    """Schema for updating an EnvironmentConfig"""
    config_data: Optional[Dict[str, Any]] = None
    overrides: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None


class EnvironmentConfigResponse(EnvironmentConfigBase):
    """Schema for EnvironmentConfig response"""
    id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ============== Configuration Schemas ==============

class ConfigurationBase(BaseModel):
    """Base schema for Configuration"""
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    config_schema_id: UUID


class ConfigurationCreate(ConfigurationBase):
    """Schema for creating a new Configuration"""
    created_by: Optional[str] = None


class ConfigurationUpdate(BaseModel):
    """Schema for updating a Configuration"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    current_version_id: Optional[UUID] = None
    is_active: Optional[bool] = None


class ConfigurationResponse(ConfigurationBase):
    """Schema for Configuration response"""
    id: UUID
    current_version_id: Optional[UUID]
    is_active: bool
    created_by: Optional[str]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# ============== Validation Schemas ==============

class ConfigValidationRequest(BaseModel):
    """Request schema for config validation"""
    config_data: Dict[str, Any]
    schema_id: UUID


class ValidationError(BaseModel):
    """Schema for individual validation errors"""
    path: str
    message: str
    error_type: str


class ConfigValidationResponse(BaseModel):
    """Response schema for config validation"""
    is_valid: bool
    errors: List[ValidationError] = []
    schema_id: UUID
    validated_at: datetime = Field(default_factory=datetime.utcnow)


# ============== Diff Schemas ==============

class DiffEntry(BaseModel):
    """Schema for a single diff entry"""
    path: str
    old_value: Any
    new_value: Any
    change_type: str  # ADDED, REMOVED, MODIFIED


class ConfigDiffRequest(BaseModel):
    """Request schema for config diff"""
    config_a: Dict[str, Any]
    config_b: Dict[str, Any]


class ConfigDiffResponse(BaseModel):
    """Response schema for config diff"""
    has_differences: bool
    differences: List[DiffEntry] = []
    added_count: int = 0
    removed_count: int = 0
    modified_count: int = 0
    compared_at: datetime = Field(default_factory=datetime.utcnow)


# ============== Rollback Schemas ==============

class RollbackRequest(BaseModel):
    """Request schema for rollback operation"""
    version_id: UUID
    changelog: Optional[str] = "Rollback to previous version"


class RollbackResponse(BaseModel):
    """Response schema for rollback operation"""
    success: bool
    previous_version: Optional[ConfigVersionResponse]
    new_version: Optional[ConfigVersionResponse]
    message: str


# ============== Template Application Schemas ==============

class ApplyTemplateRequest(BaseModel):
    """Request schema for applying a template"""
    template_id: UUID
    parameters: Dict[str, Any] = {}


class ApplyTemplateResponse(BaseModel):
    """Response schema for template application"""
    success: bool
    applied_config: Dict[str, Any]
    template_id: UUID
    applied_at: datetime = Field(default_factory=datetime.utcnow)


# ============== List/Paginated Response Schemas ==============

class PaginatedConfigSchemaResponse(BaseModel):
    """Paginated response for config schemas"""
    items: List[ConfigSchemaResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class PaginatedConfigVersionResponse(BaseModel):
    """Paginated response for config versions"""
    items: List[ConfigVersionResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class PaginatedConfigTemplateResponse(BaseModel):
    """Paginated response for config templates"""
    items: List[ConfigTemplateResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class PaginatedConfigSnapshotResponse(BaseModel):
    """Paginated response for config snapshots"""
    items: List[ConfigSnapshotResponse]
    total: int
    page: int
    page_size: int
    has_more: bool
